src/main/java/com/yahoo/memory/WritableBufferImpl.java

```diff
@@ -51,66 +51,82 @@
 
   static {
-    ZERO_SIZE_BUFFER = new WritableBufferImpl(new ResourceState(new byte[0], Prim.BYTE, 0));
+    ZERO_SIZE_BUFFER = new WritableBufferImpl(new ResourceState(new byte[0], Prim.BYTE, 0), true);
     ZERO_SIZE_BUFFER.originMemory = WritableMemoryImpl.ZERO_SIZE_MEMORY;
   }
 
-  WritableBufferImpl(final ResourceState state) {
-    super(state);
+  WritableBufferImpl(final ResourceState state, final boolean localReadOnly) {
+    super(state, localReadOnly);
   }
 
+  //  static WritableBufferImpl newInstance(final ResourceState state, final boolean localReadOnly) {
+  //    if (state.getCapacity() == 0) { return ZERO_SIZE_BUFFER; }
+  //    return new WritableBufferImpl(state, localReadOnly);
+  //  }
+
   //DUPLICATES & REGIONS XXX
   @Override
   public Buffer duplicate() {
-    final Buffer duplicate = writableDuplicate();
-    duplicate.getResourceState().setResourceReadOnly();
-    return duplicate;
+    return writableDuplicateImpl(false);
   }
 
   @Override
   public WritableBuffer writableDuplicate() {
+    return writableDuplicateImpl(localReadOnly);
+  }
+
+  private WritableBuffer writableDuplicateImpl(final boolean localReadOnly) {
     checkValid();
     if (capacity == 0) { return ZERO_SIZE_BUFFER; }
-    final WritableBufferImpl wBufImpl = new WritableBufferImpl(state);
+    final WritableBufferImpl wBufImpl = new WritableBufferImpl(state, localReadOnly);
     wBufImpl.setStartPositionEnd(getStart(), getPosition(), getEnd());
     return wBufImpl;
   }
 
   @Override
   public Buffer region() {
-    final Buffer region = writableRegion(getPosition(), getEnd() - getPosition());
-    region.getResourceState().setResourceReadOnly();
-    return region;
+    return writableRegionImpl(getPosition(), getEnd() - getPosition(), true);
+
   }
 
   @Override
   public WritableBuffer writableRegion() {
-    return writableRegion(getPosition(), getEnd() - getPosition());
+    return writableRegionImpl(getPosition(), getEnd() - getPosition(), false);
   }
 
   @Override
   public WritableBuffer writableRegion(final long offsetBytes, final long capacityBytes) {
+    return writableRegionImpl(offsetBytes, capacityBytes, localReadOnly);
+  }
+
+  private WritableBuffer writableRegionImpl(final long offsetBytes, final long capacityBytes,
+      final boolean localReadOnly) {
     checkValidAndBounds(offsetBytes, capacityBytes);
     if (capacityBytes == 0) { return ZERO_SIZE_BUFFER; }
     final ResourceState newState = state.copy();
     newState.putRegionOffset(newState.getRegionOffset() + offsetBytes);
     newState.putCapacity(capacityBytes);
-    final WritableBufferImpl wBufImpl = new WritableBufferImpl(newState);
+    final WritableBufferImpl wBufImpl = new WritableBufferImpl(newState, localReadOnly);
     wBufImpl.setStartPositionEnd(0L, 0L, capacityBytes);
     return wBufImpl;
+
   }
 
   //MEMORY XXX
   @Override
   public Memory asMemory() {
-    return asWritableMemory();
+    return asWritableMemoryImpl(true);
```

Comment:
If on *writable* buffer `asMemory()` is called first, it will "install" read-only `originMemory`. And then, when `asWritableMemory()` is called, it will return not writable `WritableMemory`, that is wrong.

Reply:
You are right!  I think it needs to be changed to this:

```java
  private WritableMemory asWritableMemoryImpl(final boolean localReadOnly) {
    checkValid();
    if ((originMemory != null) && (originMemory.localReadOnly == localReadOnly)) {
      return originMemory;
    }
    return new WritableMemoryImpl(state, localReadOnly);
  }
```